client/my-sites/email/paths.ts

```diff
@@ -0,0 +1,189 @@
+import { stringify } from 'qs';
+import { isUnderDomainManagementAll, domainManagementRoot } from 'calypso/my-sites/domains/paths';
+
+type QueryStringParameters = { [ key: string ]: string | undefined };
+
+type EmailPathUtilityFunction = (
+	siteName: string | null | undefined,
+	domainName?: string | null,
+	relativeTo?: string | null,
+	urlParameters?: QueryStringParameters
+) => string;
+
+export const emailManagementPrefix = '/email';
+export const emailManagementAllSitesPrefix = '/email/all';
+
+export function isUnderEmailManagementAll( path?: string | null ) {
+	return path?.startsWith( emailManagementAllSitesPrefix + '/' );
+}
+
+// Builds a URL query string from an object. Handles null values.
+const buildQueryString = ( parameters?: QueryStringParameters ) =>
+	parameters ? stringify( parameters, { addQueryPrefix: true, skipNulls: true } ) : '';
+
+function resolveRootPath( relativeTo?: string | null ) {
+	if ( relativeTo === emailManagementAllSitesPrefix || relativeTo === domainManagementRoot() ) {
+		return emailManagementAllSitesPrefix;
+	}
+
+	if ( isUnderEmailManagementAll( relativeTo ) || isUnderDomainManagementAll( relativeTo ) ) {
+		return emailManagementAllSitesPrefix;
+	}
+
+	return emailManagementPrefix;
+}
+
+function getPath(
+	siteName: string | null | undefined,
+	domainName: string | null | undefined,
+	slug: string,
+	relativeTo?: string | null,
+	urlParameters?: QueryStringParameters
+) {
+	if ( siteName && domainName ) {
+		// Encodes only real domain names and not parameter placeholders
+		if ( ! domainName.startsWith( ':' ) ) {
+			// Encodes domain names so addresses with slashes in the path (e.g. used in site redirects) don't break routing.
+			// Note they are encoded twice since page.js decodes the path by default.
+			domainName = encodeURIComponent( encodeURIComponent( domainName ) );
+		}
+
+		return (
+			resolveRootPath( relativeTo ) +
+			'/' +
+			domainName +
+			'/' +
+			slug +
+			'/' +
+			siteName +
+			buildQueryString( urlParameters )
+		);
+	}
+
+	if ( siteName ) {
+		return '/email/' + siteName + buildQueryString( urlParameters );
```

Comment:
Don't you want to also call `resolveRootPath( relativeTo )` here? The original `emailManagementEdit` function did it.

Reply:
The original `emailManagementEdit` function would indeed call `resolveRootPath` if `domainName` was nullish, but it also wasn't really set up to return a working path in those cases. This is illustrated in the unit tests that I just landed:
https://github.com/Automattic/wp-calypso/blob/0eed03aa7e4d0772f2ac0af62500d8128ae43409/client/my-sites/email/test/paths.ts#L74-L76

With this refactoring, I've opted to always return a working path (even when `siteName` or `domainName` are nullish). The fallback path here is `/email/:site`. If we were to call `resolveRootPath` here, we'd sometimes return `/email/all/:site`, which isn't actually a path supported by `email/index.js`:
https://github.com/Automattic/wp-calypso/blob/0eed03aa7e4d0772f2ac0af62500d8128ae43409/client/my-sites/email/index.js#L33-L59